### DISCUSSION TITLE
RFC: Don't use Numbers for Content-File links #RFC2021-02

### DISCUSSION THREAD

**Discussion by iJungleboy** [oqtane/oqtane.framework#1292] [General]
As of now, content-files use a number to download them. This has major issues

1. It's ugly
2. It's impossible to guess what this file will be (pdf, img, dangerous JS)
3. Someone can just start going through numbers and see what he can get - which I perceive as a security issue

IMHO the mechanism should be 

`[some-root-path]/[optional-sub-paths]/filename.ext`

It does add a trivial extra load to the DB, but I really feel that the numbers must die - especially because of the security risk that people can just start downloading by-number.

**Comment by sbwalker** [comment]
There is no security risk as the download of files is secured on the server based on permissions. So although it is true that a user can iterate through IDs in API requests there is no risk in file content being exposed to unauthorized users.

**Reply by iJungleboy** [reply]
In theory I agree but not in practice 😉. From my experience people rarely spend the time to "do this right" and not using IDs would increase protection for the 99% of people who didn't turn on their thinking caps when creating something. 

But I understand that this would be a breaking change and need some consideration.

**Comment by sbwalker** [comment]
Utilizing filenames rather than IDs is essentially relying on "security by obscurity" - substituting real security for secrecy/obfuscation - which is an anti-pattern and not an effective security practice. Systems need to rely on real security measures in order to be safe... and administrators need to utilize these security measures so they do not leave themselves vulnerable.

**Comment by JustCallMeAD** [comment]
> 3\. Someone can just start going through numbers and see what he can get - which I perceive as a security issue

@iJungleboy I agree it is a security issue and previosly mentioned something similar in relation to user account information https://github.com/oqtane/oqtane.framework/issues/645

> which is an anti-pattern and not an effective security practice

@sbwalker Paler's data was scaped because they used sequential numbers for their posts:
https://techbeacon.com/security/scraped-parler-data-reveals-countless-capitol-perps

> Parler lacked the most basic security measures that would have prevented the automated scraping. … It even ordered its posts by number in the site's URLs, so that anyone could have easily, programmatically downloaded the site's millions of posts [a] cardinal security sin … known as an insecure direct object reference [or] IDOR. … Services like Twitter, by contrast, randomize the URLs of posts so they can't be guessed.

Making defualt url's, id's or usernames easily guessable is certainly security issue.

There are [many](https://www.clever-cloud.com/blog/engineering/2015/05/20/why-auto-increment-is-a-terrible-idea/), [many](https://wordpress.org/support/topic/change-the-admin-user-name/) more examples on the web that prove using sequential numbers as url's, id's and fixed usernames such as "host" and "admin" are bad practices. Unless you have a specific need to use sequential numbers, randomized id's / url's should always be used. 

This are relatively easy issues to fix now and will only get harder to fix in the future when Oqtane on boards more users.

**Reply by sbwalker** [reply]
**Just to be 100% clear for people reading this thread**... the existing approach for content files in Oqtane is secure. There is no way for an unauthorized user to download files as long as the Folder permissions related to the file have been set correctly by the administrator. Oqtane does not rely on "security by obscurity" - it relies on granular permissions for all content assets. Suggesting that Oqtane has a security issue because it uses IDs in Urls is incorrect and misleading.

**Reply by JustCallMeAD** [reply]
@sbwalker thank you again for all the work you have done on oqtane as you have done a great job and it is much appreciated.

The suggestions were how to make oqtane more secure. There are a lot of lessons to be learned from other projects which have had known and common flaws in the past, especially when it comes to security as easily guessable id's and usernames has been one of the issues that has been a common reoccurrence out there in the wild.

> There is no way for an unauthorized user to download files as long as the Folder permissions related to the file have been set correctly by the administrator.

In a perfect world that may be true. All you need is someone to accidentally use the wrong permissions on a page, wrong permissions on a module or some badly written code that a bad actor or hacker can take advantage of. I think that we can all agree that taking years instead of seconds to guess something is a lot more secure. 

Example:

fileid=1
(I bet in 1 second anyone can guess what the next fileid is going to be)

fileid=8d9f6679-7425-94de-944b-e07fc1f90ea8
(I bet that even if I gave someone 1 year to guess the next fileid that they could not do it - the odds are not in their favor)

https://stackoverflow.com/questions/4878359/what-is-the-probability-of-guessing-matching-a-guid

"So, even if you generate 2^60 GUIDs, the odds of a collision are extremely small. If you can generate one billion GUIDs per second, it would still take 36 years to have a 1.95e-03 chance of a collision."

> Oqtane does not rely on "security by obscurity"

The suggestion in this thread was not to just rely on "security by obscurity" as having granular permissions is important too, but having some "security by obscurity" is better than "no security by obscurity". Making something more secure is a good thing, not a bad thing - you can never be too secure. 

Also, as I stated in my other post, "information leakage" such as other people knowing how many files or users you have can be a concern to some people, especially businesses.

Probably more related to my other post but worth thinking about is exporting / importing data. GUIDs make this easier as there is much less chance of collision and may help when users importing / exporting site and module content.

I hope that you are open to continue to discuss issues like this as at the end of the day we are all just trying to help build a better product.

**Comment by JustCallMeAD** [comment]
@iJungleboy your suggestion of "[some-root-path]/[optional-sub-paths]/filename.ext" is a good idea. How about if the filename was not unique or the filename itself contained sensitive information?

**Reply by iJungleboy** [reply]
@JustCallMeAD the file name must be unique, otherwise it couldn't be stored in the file system :)

The filename could of course contain something sensitive but to be honest that's not a common theme. Let's assume people posted their bio on a platform then whoever gets to a page with the link is very likely also allowed to see the file - since that's what developers usually ensure. 

My concern is that the developer usually has a much harder time securing the files in complex scenarios, and often will forget about it in the end. In that case I as of now, a user could just skim the numbers and get each file. 

We're currently developing something to be released on Tuesday, and we actually had to put our files in a different location because this was just too risky.

**Reply by JustCallMeAD** [reply]
> the file name must be unique, otherwise it couldn't be stored in the file system :)

Good point..lol. That is fine for most situations as I forgot that you need to check to see if the file exists anyway.

Sometimes we use [Path.GetRandomFileName()](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getrandomfilename?view=net-5.0) to generate a random filename on disk and then just rewrite the file name on download.

**Comment by sbwalker** [comment]
To provide some additional context... the File/Folder API was developed specifically with security in mind. It is an abstraction over the physical file system so that an extra layer of security exists between the users and the assets. The files themselves are physically stored in an inaccessible folder location outside of wwwroot so that there is no possible way that they can be accessed except through the application itself. Folders have granular security where it is possible to specify the user/roles who can browse the folder contents, view a file, or add/edit files in a folder. If a user which does not have the necessary permissions tries to access a file, they will be blocked. This is a robust security scheme with a proven track record in enterprise environments. 

One of the fundamental reasons a secure File/Folder API was included in Oqtane was because one of the more common security vulnerabilities which occurred in DNN over the years was related to third party developers rolling their own scheme for managing/uploading files. There were numerous instances where vulnerabilities were exposed because these custom file mechanisms were either unprotected or reliant upon weak security mechanisms. Often these vulnerabilities were a result of the third party developer not fully understanding the details of multi-tenancy or other aspects of the framework. In addition they usually relied on a different permissions management user experience than the framework itself, which resulted in confusion on the part of the user to understand the specifics of various approaches. 

In regards to the use of IDs, I will say that using names or GUIDs can add an additional layer of protection through "security by obscurity"  - but it is not a real security approach on its own. Essentially it attempts to make it difficult to guess the entry point to your system - but if a user is able make a correct guess then they would have full access to the system. There is no enterprise which would approve this approach as "security"... however it could be layered on top of a real security system to make it more resilient. 

That being said, backward compatibility is one of Oqtane's core philosophies so I am not in favor of introducing a breaking change to an area of the framework which is already 100% secure.
